core/node/libp2p/routing.go

```diff
@@ -32,23 +35,89 @@ type p2pRouterOut struct {
 	Router Router `group:"routers"`
 }
 
-func BaseRouting(lc fx.Lifecycle, in BaseIpfsRouting) (out p2pRouterOut, dr *ddht.DHT) {
-	if dht, ok := in.(*ddht.DHT); ok {
-		dr = dht
+type processInitialRoutingIn struct {
+	fx.In
+
+	Router routing.Routing `name:"initialrouting"`
+
+	// For setting up experimental DHT client
+	Host      host.Host
+	Repo      repo.Repo
+	Validator record.Validator
+}
+
+type processInitialRoutingOut struct {
+	fx.Out
+
+	Router    Router `group:"routers"`
+	DHT       *ddht.DHT
+	DHTClient routing.Routing `name:"dhtc"`
+	BaseRT    BaseIpfsRouting
+}
 
-		lc.Append(fx.Hook{
-			OnStop: func(ctx context.Context) error {
-				return dr.Close()
+func BaseRouting(experimentalDHTClient bool) interface{} {
+	return func(mctx helpers.MetricsCtx, lc fx.Lifecycle, in processInitialRoutingIn) (out processInitialRoutingOut, err error) {
+		var dr *ddht.DHT
+		if dht, ok := in.Router.(*ddht.DHT); ok {
+			dr = dht
+
+			lc.Append(fx.Hook{
+				OnStop: func(ctx context.Context) error {
+					return dr.Close()
```

Comment:
This is strange... I'd expect DI function constructing the DHT to register a close hook.

Reply:
yep, although this was here before so I'm going to leave it for now. I suspect it's related to the initial routing being set specially via a variable in the build config object.